### PR TITLE
Allow disabling of client type via allowed_clients param

### PR DESCRIPTION
**Pull request by chatton**
## Description

This PR directly replaces #3194

closes: #2537

### Commit Message / Changelog Entry

N/A included in feature branch

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/02-client/keeper/keeper.go [hunk 1]
```diff
@@ -403,3 +403,12 @@ func (k Keeper) ClientStore(ctx sdk.Context, clientID string) sdk.KVStore {
 	clientPrefix := []byte(fmt.Sprintf("%s/%s/", host.KeyClientStorePrefix, clientID))
 	return prefix.NewStore(ctx.KVStore(k.storeKey), clientPrefix)
 }
+
+// GetClientStatus returns the status for a given clientState. If the client type is not in the allowed
+// clients param field, Unauthorized is returned, otherwise the client state status is returned.
+func (k Keeper) GetClientStatus(ctx sdk.Context, clientState exported.ClientState, clientID string) exported.Status {
```
Comment (colin-axner) [inline modules/core/02-client/keeper/keeper.go:409]: Should we add documentation somewhere that removing clients from the `AllowedClients` parameter will disable all usage of that client type?

**Review by colin-axner — APPROVED** [review]
This is wonderful! So fast! We should probably add a changelog entry for this and I think it would be good to add a little docs on this functionality. It might be unclear when to use `clientState.Status()` vs `clientKeeper.GetClientStatus`

File: modules/core/02-client/keeper/keeper.go (hunk 1 above)
Comment (chatton) [reply modules/core/02-client/keeper/keeper.go:409]: Yes I think we can improve this docstring a bit!

**Review by damiannolan — APPROVED** [review]
Beautiful, super clean PR! LGTM 🙌

File: docs/ibc/light-clients/client-state.md [hunk 2]
```diff
@@ -28,6 +28,8 @@ implementer is in charge of determining which checks are required. See the [Tend
 - A `Frozen` status indicates that misbehaviour was detected in the counterparty chain and the client is not allowed to be used.
 - An `Expired` status indicates that a client is not allowed to be used because it was not updated for longer than the trusting period.
 - An `Unknown` status indicates that there was an error in determining the status of a client.
+- An `Unauthorized` status indicates that usage of the client is not allowed.
+  - > Note: Removing a client type from the [`AllowedClients`](https://github.com/cosmos/ibc-go/blob/v6.0.0/modules/core/02-client/types/client.pb.go#L345) array will prevent all usage of that client type.
```
Comment (chatton) [inline docs/ibc/light-clients/client-state.md:32]: Added a note in the client-state.md doc. 

cc @colin-axner

Comment (colin-axner) [reply docs/ibc/light-clients/client-state.md]: Hmm, the only issue with the doc here is that this is for the `clientState.Status` message. Light client developers don't need to use the `Unauthorized` status

Comment (chatton) [reply docs/ibc/light-clients/client-state.md]: That is a very good point! Let me see if I can find a better place for this.

File: docs/ibc/light-clients/client-state.md [hunk 3]
```diff
@@ -28,6 +28,8 @@ implementer is in charge of determining which checks are required. See the [Tend
 - A `Frozen` status indicates that misbehaviour was detected in the counterparty chain and the client is not allowed to be used.
 - An `Expired` status indicates that a client is not allowed to be used because it was not updated for longer than the trusting period.
 - An `Unknown` status indicates that there was an error in determining the status of a client.
+- An `Unauthorized` status indicates that usage of the client is not allowed.
```
Comment (crodriguezvega) [inline docs/ibc/light-clients/client-state.md:31]: The formatting of the rendered text looks a bit weird with the `- >`. Was its use intentional?

**Review by crodriguezvega — COMMENTED** [review]
Nice work, @chatton!

Should we add some tests to exercise the new conditional checks added to `ChanOpenInit` and `ConnOpenInit`?

**Comment by damiannolan** [comment]
Could we add the status check before updating the localhost client in `BeginBlocker` in this PR?

File: docs/ibc/light-clients/client-state.md (hunk 3 above)
Comment (chatton) [reply docs/ibc/light-clients/client-state.md:31]: this was intentional to add a `note` section, but I think Colin brought up a good point that this is not the best place for these docs.

File: modules/core/04-channel/keeper/handshake.go [hunk 4]
```diff
@@ -50,6 +51,15 @@ func (k Keeper) ChanOpenInit(
 		)
 	}
 
+	clientState, found := k.clientKeeper.GetClientState(ctx, connectionEnd.ClientId)
+	if !found {
+		return "", nil, sdkerrors.Wrapf(clienttypes.ErrClientNotFound, "clientID (%s)", connectionEnd.ClientId)
+	}
+
+	if status := k.clientKeeper.GetClientStatus(ctx, clientState, connectionEnd.ClientId); status != exported.Active {
```
Comment (AdityaSripal) [inline modules/core/04-channel/keeper/handshake.go:59]: Do we need this for ChanCloseInit as well?

**Review by AdityaSripal — COMMENTED** [review]
I think we need a check on ChanCloseInit as well correct?

Also does the current test coverage cover the new k.ClientStatus function completely? Would be good to ensure we get an error if we try to call a method on a client that has a valid internal status but is not part of the AllowedClients params

**Comment by chatton** [comment]
Based on PR feedback, I've 
* added a new Troubleshooting section to the docs
* added test cases which explicitly remove the tendermint client from the allow list
* added the check to `ChanCloseInit`
* added the check in `BeginBlocker`

**Review by AdityaSripal — APPROVED** [review]
Thanks for the quick fixes @chatton !

**Review by colin-axner — APPROVED** [review]
Excellent!

**Review by damiannolan — APPROVED** [review]

**Pull request merged**